src/engine/AbsoluteValueConstraint.cpp

```diff
@@ -0,0 +1,454 @@
+    fixPhaseIfNeeded();
+
+    // Update partner's bound
+    if ( isActive() && _constraintBoundTightener )
+    {
+        if ( variable == _b )
+        {
+            if ( bound > 0 )
+            {
+                double fUpperBound = FloatUtils::max( bound, -_lowerBounds[_b] );
+                _constraintBoundTightener->registerTighterUpperBound( _f, fUpperBound );
+            }
+            else
+            {
+                // Phase is fixed, don't care about this case
+            }
+        }
+        else
+        {
+            // F's upper bound can restrict both bounds of B
+            if ( bound < _upperBounds[_b] )
+                _constraintBoundTightener->registerTighterUpperBound( _b, bound );
+
+            if ( -bound > _lowerBounds[_b] )
+                _constraintBoundTightener->registerTighterLowerBound( _b, -bound );
+        }
+    }
+}
+
+bool AbsoluteValueConstraint::participatingVariable(unsigned variable ) const
+{
+    return ( variable == _b ) || ( variable == _f );
+}
+
+List<unsigned> AbsoluteValueConstraint::getParticipatingVariables() const
+{
+    return List<unsigned>( { _b, _f } );
+}
+
+bool AbsoluteValueConstraint::satisfied() const
+{
+    if ( !( _assignment.exists( _b ) && _assignment.exists( _f ) ) )
+        throw MarabouError( MarabouError::PARTICIPATING_VARIABLES_ABSENT );
+
+    double bValue = _assignment.get( _b );
+    double fValue = _assignment.get( _f );
+
+    // Possible violations:
+    //   1. f is negative
+    //   2. f is positive, abs(b) and f are not equal
+
+    if ( fValue < 0 )
+        return false;
+
+    return FloatUtils::areEqual( FloatUtils::abs( bValue ),
+                                 fValue,
+                                 GlobalConfiguration::ABS_CONSTRAINT_COMPARISON_TOLERANCE );
+}
+
+List<PiecewiseLinearConstraint::Fix> AbsoluteValueConstraint::getPossibleFixes() const
+{
+    ASSERT( !satisfied() );
+    ASSERT( _assignment.exists( _b ) );
+    ASSERT( _assignment.exists( _f ) );
+
+    double bValue = _assignment.get( _b );
+    double fValue = _assignment.get( _f );
+
+    ASSERT( !FloatUtils::isNegative( fValue ) );
+
+    List<PiecewiseLinearConstraint::Fix> fixes;
+
+    // Possible violations:
+    //   1. f is positive, b is positive, b and f are unequal
+    //   2. f is positive, b is negative, -b and f are unequal
+
+    fixes.append( PiecewiseLinearConstraint::Fix( _b, fValue ) );
+    fixes.append( PiecewiseLinearConstraint::Fix( _b, -fValue ) );
+    fixes.append( PiecewiseLinearConstraint::Fix( _f, FloatUtils::abs( bValue ) ) );
+
+    return fixes;
+}
+
+List<PiecewiseLinearConstraint::Fix> AbsoluteValueConstraint::getSmartFixes( ITableau */* tableau */ ) const
+{
+    return getPossibleFixes();
+}
+
+List<PiecewiseLinearCaseSplit> AbsoluteValueConstraint::getCaseSplits() const
+{
+    if ( _phaseStatus != PhaseStatus::PHASE_NOT_FIXED )
+        throw MarabouError( MarabouError::REQUESTED_CASE_SPLITS_FROM_FIXED_CONSTRAINT );
```

Comment:
Is this an error we should to recover from? Should we just assert this as an invariant?

Reply:
Line 268: uses the assert form for the same check.

Reply:
Fixed